src/mpc-hc/MediaTransControls.cpp

```diff
@@ -0,0 +1,234 @@
+        this->controls->put_PlaybackStatus(MediaPlaybackStatus::MediaPlaybackStatus_Playing);
+    }
+}
+
+template <typename T>
+bool AwaitForIAsyncOperation(CComPtr<ABI::Windows::Foundation::IAsyncOperation<T>> io) {
+    CComPtr<ABI::Windows::Foundation::IAsyncInfo> info;
+    ABI::Windows::Foundation::AsyncStatus status;
+    HRESULT ret;
+    info = io;
+    while (true) {
+        if ((ret = info->get_Status(&status)) != S_OK) {
+            return false;
+        }
+        if (status != ABI::Windows::Foundation::AsyncStatus::Started) {
+            if (status == ABI::Windows::Foundation::AsyncStatus::Completed) return true;
+            return false;
+        }
+        Sleep(10);
+    }
+}
+
+void MediaTransControls::loadThumbnail(CString fn) {
+    if (fn.IsEmpty() || !updater) return;
+    if (PathUtils::IsURL(fn)) return loadThumbnailFromUrl(fn);
+    HRESULT ret;
+    CComPtr<IStorageFileStatics> sfs;
+    if ((ret = GetActivationFactory(HStringReference(RuntimeClass_Windows_Storage_StorageFile).Get(), &sfs)) != S_OK) {
+        return;
+    }
+    CComPtr<ABI::Windows::Foundation::IAsyncOperation<StorageFile*>> af;
+    if ((ret = sfs->GetFileFromPathAsync(HStringReference(fn.GetString()).Get(), &af)) != S_OK) {
+        return;
+    }
+    /// Present file
+    CComPtr<IStorageFile> f;
+    if (!AwaitForIAsyncOperation(af)) {
+        return;
+    }
+    if ((ret = af->GetResults(&f)) != S_OK) {
+        return;
+    }
+    CComPtr<Streams::IRandomAccessStreamReferenceStatics> rasrs;
+    if ((ret = GetActivationFactory(HStringReference(RuntimeClass_Windows_Storage_Streams_RandomAccessStreamReference).Get(), &rasrs)) != S_OK) {
+        return;
+    }
+    CComPtr<Streams::IRandomAccessStreamReference> stream;
+    if ((ret = rasrs->CreateFromFile(f, &stream)) != S_OK) {
+        return;
+    }
+    updater->put_Thumbnail(stream);
+}
+
+void MediaTransControls::loadThumbnail(BYTE* content, size_t size) {
+    if (!content || !size || !updater) return;
+    ComPtr<Streams::IRandomAccessStream> s;
+    HRESULT ret;
+    if ((ret = ActivateInstance(HStringReference(RuntimeClass_Windows_Storage_Streams_InMemoryRandomAccessStream).Get(), s.GetAddressOf())) != S_OK) {
+        return;
+    }
+    ComPtr<IStream> writer;
+    CreateStreamOverRandomAccessStream(s.Get(), IID_PPV_ARGS(writer.GetAddressOf()));
+    writer->Write(content, size, nullptr);
+    CComPtr<Streams::IRandomAccessStreamReferenceStatics> rasrs;
+    if ((ret = GetActivationFactory(HStringReference(RuntimeClass_Windows_Storage_Streams_RandomAccessStreamReference).Get(), &rasrs)) != S_OK) {
+        return;
+    }
+    CComPtr<Streams::IRandomAccessStreamReference> stream;
+    if ((ret = rasrs->CreateFromStream(s.Get(), &stream)) != S_OK) {
+        return;
+    }
+    updater->put_Thumbnail(stream);
+}
+
+void MediaTransControls::loadThumbnailFromUrl(CString url) {
+    if (url.IsEmpty() || !updater) return;
+    if (PathUtils::IsFile(url)) return loadThumbnail(url);
```

Comment:
I would prefer to remove this line. Just to prevent possible infinite recursion in case of bug ever happens in PathUtils::IsFile().

loadThumbnailFromUrl() is only called with URL, so no need to check the input parameter again.